vslib/src/sai_vs_hostintf.cpp

```diff
@@ -29,7 +67,414 @@ sai_status_t vs_send_hostif_packet(
+
+    SWSS_LOG_NOTICE("setup forward rule for %s succeeded", tapname.c_str());
+
+    return true;
+}
+
+sai_status_t vs_create_hostif_int(
+        _In_ sai_object_type_t object_type,
+        _Out_ sai_object_id_t *hostif_id,
+        _In_ sai_object_id_t switch_id,
+        _In_ uint32_t attr_count,
+        _In_ const sai_attribute_t *attr_list)
+{
+    SWSS_LOG_ENTER();
+
+    if (g_vs_hostif_use_tap_device == false)
+    {
+        return SAI_STATUS_SUCCESS;
+    }
+
+    // validate SAI_HOSTIF_ATTR_TYPE
+
+    auto attr_type = sai_metadata_get_attr_by_id(SAI_HOSTIF_ATTR_TYPE, attr_count, attr_list);
+
+    if (attr_type == NULL)
+    {
+        SWSS_LOG_ERROR("attr SAI_HOSTIF_ATTR_TYPE was not passed");
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (attr_type->value.s32 != SAI_HOSTIF_TYPE_NETDEV)
+    {
+        SWSS_LOG_ERROR("only SAI_HOSTIF_TYPE_NETDEV is supported");
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    // validate SAI_HOSTIF_ATTR_OBJ_ID
+
+    auto attr_obj_id = sai_metadata_get_attr_by_id(SAI_HOSTIF_ATTR_OBJ_ID, attr_count, attr_list);
+
+    if (attr_obj_id == NULL)
+    {
+        SWSS_LOG_ERROR("attr SAI_HOSTIF_ATTR_OBJ_ID was not passed");
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    sai_object_id_t obj_id = attr_obj_id->value.oid;
+
+    sai_object_type_t ot = sai_object_type_query(obj_id);
+
+    if (ot != SAI_OBJECT_TYPE_PORT)
+    {
+        SWSS_LOG_ERROR("SAI_HOSTIF_ATTR_OBJ_ID=%s expected to be PORT but is: %s",
+                sai_serialize_object_id(obj_id).c_str(),
+                sai_serialize_object_type(ot).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    // validate SAI_HOSTIF_ATTR_NAME
+
+    auto attr_name = sai_metadata_get_attr_by_id(SAI_HOSTIF_ATTR_NAME, attr_count, attr_list);
+
+    if (attr_name == NULL)
+    {
+        SWSS_LOG_ERROR("attr SAI_HOSTIF_ATTR_NAME was not passed");
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (strnlen(attr_name->value.chardata, sizeof(attr_name->value.chardata)) >= MAX_INTERFACE_NAME_LEN)
+    {
+        SWSS_LOG_ERROR("interface name is too long: %.*s", MAX_INTERFACE_NAME_LEN, attr_name->value.chardata);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (strncmp(attr_name->value.chardata, "Ethernet", 8) != 0)
+    {
+        SWSS_LOG_ERROR("interface name should start with EthernetX but is %s", attr_name->value.chardata);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    std::string name = std::string(attr_name->value.chardata);
+
+    // create TAP device
+
+    SWSS_LOG_INFO("creating hostif %s", name.c_str());
+
+    int tapfd = vs_create_tap_device(name.c_str(), IFF_TAP | IFF_MULTI_QUEUE | IFF_NO_PI);
+
+    if (tapfd < 0)
+    {
+        SWSS_LOG_ERROR("failed to create TAP device for %s", name.c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    SWSS_LOG_INFO("created TAP device for %s, fd: %d", name.c_str(), tapfd);
+
+    // XXX currently tapfd is ignored, it should be closed on hostif_remove
```

Comment:
>XXX [](start = 7, length = 3)

XXX is confusing. Use TODO?

Reply:
XXX is used to get programmers attention, but i will change to TODO

Reply:
fixed